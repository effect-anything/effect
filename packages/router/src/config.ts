import { EventEmitter } from "events"
import { RouteConfig } from "./types"

export class RouteConfigList extends EventEmitter {
  private _routes: RouteConfig[] = []

  get routes() {
    return this._routes
  }

  constructor(routes: RouteConfig[]) {
    super()

    this._routes = routes
  }

  public combine(path: string, routeConfig: RouteConfig) {
    const index = this._routes.findIndex((x) => x.path === path)

    if (index === -1) {
      return
    }

    this._routes[index].children?.push(routeConfig)

    this.update()

    return this
  }

  public add(routeConfig: RouteConfig) {
    this._routes.push(routeConfig)

    this.update()

    return this
  }

  public update() {
    this.emit("update")
  }
}
