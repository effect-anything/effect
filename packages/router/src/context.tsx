import { createContext, useContext } from "react"
import type { RouteNode } from "./node"

export interface RouterContextState {
  routes: RouteNode[]
  elements: JSX.Element | null
}

const createNamedContext = (name: string) => {
  const context = createContext<RouterContextState>({
    routes: [],
    elements: null,
  })

  context.displayName = name

  return context
}

export const context = /* #__PURE__ */ createNamedContext("RettRouterContext")

export const useRoutes = (): RouteNode[] => {
  return useContext(context).routes
}

export const useRoutesElement = (): JSX.Element | null => {
  return useContext(context).elements
}
