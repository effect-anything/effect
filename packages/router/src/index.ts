export type { PipeFn, RouteComponent, RouteConfig, RouteLocation, RouteTagType } from "./types"
export * from "./node"
export * from "./helper"
export * from "./context"
export * from "./config"
export * from "./breadcrumb/match"
