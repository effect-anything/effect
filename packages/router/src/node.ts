import { any, clone, compose, identity, ifElse, map, reduce, reduced, reject } from "ramda"
import type { RouteConfig } from "./types"
import { getRouteProps, isRealEmpty, joinPath } from "./utils"

type RouteNodeMeta = RouteConfig["meta"] & {
  /**
   * 原路径
   */
  readonly _path: string

  /**
   * 自动重定向地址
   */
  readonly autoRedirectPath?: string
}

/**
 * 过滤不符合规则配置
 **/
export const rejectIllegalConfig = reject<RouteConfig, "array">((x) => isRealEmpty(x.path) || !x.path)

const defaultRejectChars = ["*", ":", "?", "+", ".", "(", ")", "^", "&"]

export const filterByChars = (path: string, chars: string[] = defaultRejectChars) => {
  return any<string>((char) => path.indexOf(char) > -1, chars)
}

export const filterChildrenByParams = (routesNode?: RouteNode[]): RouteNode[] => {
  if (isRealEmpty(routesNode)) {
    return []
  }

  return routesNode!.filter((routeConfig) => {
    const charHit = filterByChars(routeConfig.path)

    return !charHit
  })
}

export const getRouteNodeParents = (routeNode?: RouteNode, results: RouteNode[] = []): RouteNode[] => {
  return routeNode ? getRouteNodeParents(routeNode.parent, results.concat(routeNode)) : results
}

export const calcAutoRedirectPath = (routesNode?: RouteNode[]): string | undefined => {
  const findPath = reduce(
    (acc, elem) => {
      if (elem.meta.autoRedirectPath) {
        return reduced(elem.meta.autoRedirectPath)
      }

      if (!elem.children) {
        return reduced(elem.path)
      }

      if (elem.hasChildren) {
        const result = calcAutoRedirectPath(elem.children)

        if (result) {
          return reduced(result)
        }

        return acc
      }

      return acc
    },
    undefined as string | undefined,
    filterChildrenByParams(routesNode)
  )

  return findPath
}

export const mapRouteNode =
  (parent: RouteNode | undefined) =>
  (routeConfig: RouteConfig[] | undefined): RouteNode[] =>
    ifElse(
      isRealEmpty,
      identity,
      compose(
        map<RouteConfig, RouteNode>((routeConfig) => {
          const cloneConfig = clone(routeConfig)

          const routeNode = new RouteNode(cloneConfig, parent)

          routeNode.updateChildren(() => mapRouteNode(routeNode)(cloneConfig.children))

          return routeNode
        }),
        rejectIllegalConfig
      )
    )(routeConfig)

export class RouteNode {
  /**
   * 路由 path
   */
  private _path: string

  /**
   * 路由组件
   */
  private _component: RouteConfig["component"]

  /**
   * 路由 Route 标签 props
   */
  private _routeProps: RouteConfig["routeProps"]

  /**
   * 路由 Meta 信息
   */
  private _meta: RouteNodeMeta

  /**
   * 子节点路由列表
   */
  private _children?: RouteNode[]

  /**
   * 子节点路由配置
   */
  private _childrenConfig?: RouteConfig[]

  /**
   * 父节点路由
   */
  private _parent?: RouteNode

  get path(): RouteConfig["path"] {
    const parentPath = this._parent?.path ?? "/"

    // 连接父节点的 path
    return joinPath(parentPath, this._path)
  }

  get routeProps(): RouteConfig["routeProps"] {
    return getRouteProps(this._routeProps)
  }

  get component(): RouteConfig["component"] {
    return this._component
  }

  get meta(): RouteNodeMeta {
    return {
      ...this._meta,
      autoRedirectPath: calcAutoRedirectPath(this._children),
    }
  }

  get children(): RouteNode[] | undefined {
    return this._children
  }

  /**
   * 父级路由节点
   */
  get parent(): RouteNode | undefined {
    return this._parent
  }

  /**
   * sibling nodes
   */
  get sibling(): RouteNode[] {
    return this.parent?.children ?? []
  }

  /**
   * 是否存在子级
   */
  get hasChildren() {
    return this._children && this._children.length > 0
  }

  constructor(routeConfig: RouteConfig, parent?: RouteNode | undefined) {
    this._path = routeConfig.path
    this._component = routeConfig.component
    this._routeProps = routeConfig.routeProps
    this._meta = {
      ...routeConfig.meta,
      _path: routeConfig.path,
    }
    this._parent = parent
    this._childrenConfig = routeConfig.children
  }

  public getParents() {
    return getRouteNodeParents(this)
  }

  public updatePath(fn: (path: RouteNode["path"]) => RouteNode["path"]) {
    this._path = fn(this._path)

    return this
  }

  public updateChildren(fn: (children: RouteNode["_children"]) => RouteNode["_children"]) {
    this._children = fn(this._children)

    return this
  }

  public updateMeta(fn: (meta: RouteNode["_meta"]) => RouteNode["_meta"]) {
    this._meta = fn(this._meta)

    return this
  }

  public clone() {
    return new RouteNode(
      clone({
        path: this._path,
        children: this._childrenConfig,
        component: this._component,
        meta: this._meta,
        routeProps: this._routeProps,
      }),
      this._parent
    ).updateChildren(() => this._children)
  }

  public toJSON(): RouteConfig {
    return {
      path: this.path,
      component: this.component,
      children: this.children?.map((x) => x.toJSON()),
      meta: this.meta,
      routeProps: this.routeProps,
    }
  }
}
