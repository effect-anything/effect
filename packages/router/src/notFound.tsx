import React from "react"
import type { FunctionComponent } from "react"
import { useLocation } from "react-router-dom"

export const DefaultNotFoundComponent: FunctionComponent = () => {
  const location = useLocation()

  return <div>404 NotFound: {location.pathname}</div>
}
