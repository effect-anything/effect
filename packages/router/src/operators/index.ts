export * from "./filter"
export * from "./map"
export * from "./reduce"
export * from "./find"
export * from "./flat"
