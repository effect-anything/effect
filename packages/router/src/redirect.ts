import type { RouteNode } from "./node"
import type { FunctionComponent } from "react"
import { useEffect } from "react"
import { useHistory, useLocation } from "react-router-dom"

interface IRouteAutoRedirect {
  routeNode: RouteNode
}

/**
 * 重定向到合适的 children 下的 path
 */
export const RouteAutoRedirect: FunctionComponent<IRouteAutoRedirect> = ({ routeNode }) => {
  const { pathname } = useLocation()
  const history = useHistory()

  useEffect(() => {
    const { autoRedirectPath } = routeNode.meta
    const pathHit = routeNode.path === pathname

    /**
     * 如果当前组件 path 不等于当前路由 pathname，此逻辑不生效
     */
    if (!pathHit) {
      return
    }

    if (!autoRedirectPath) {
      return
    }

    /**
     * 重定向登录当前配置 path 或者登录当前路由 pathname
     */
    if (autoRedirectPath === routeNode.path || autoRedirectPath === pathname) {
      return
    }

    history.replace(autoRedirectPath)
  }, [routeNode, history, pathname])

  return null
}

interface RouteRedirectProps {
  routeNode: RouteNode
}

/**
 * 路由重定向
 */
export const RouteRedirect: FunctionComponent<RouteRedirectProps> = ({ routeNode }) => {
  const { pathname } = useLocation()
  const history = useHistory()

  useEffect(() => {
    const { redirect } = routeNode.meta

    if (pathname === routeNode.path && redirect && pathname !== redirect) {
      history.push(redirect)
    }
  }, [routeNode, pathname])

  return null
}
