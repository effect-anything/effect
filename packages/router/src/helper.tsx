import pathToRegexp from "path-to-regexp"
import { compose, map, mergeDeepRight, pipe } from "ramda"
import React, { FunctionComponent, memo, useRef, useState } from "react"
import { Redirect, Route, RouteComponentProps, Switch } from "react-router-dom"
import { context as RouterContext } from "./context"
import { mapRouteNode, RouteNode } from "./node"
import { DefaultNotFoundComponent } from "./notFound"
import * as operators from "./operators"
import { RouteAutoRedirect, RouteRedirect } from "./redirect"
import { PipeFn, RouteComponent, RouteConfig, RouteLocation, RouteTagType } from "./types"
import { isRealEmpty } from "./utils"

export const findRouteNodeByPathname = (pathname: string, routeNodes: RouteNode[] | undefined) => {
  return operators.find((routeNode) => {
    if (pathname === routeNode.path) {
      return true
    }

    const match = pathToRegexp(routeNode.path, {
      sensitive: routeNode.routeProps?.sensitive,
      strict: routeNode.routeProps?.strict,
    }).exec(pathname)

    return !!match
  })(routeNodes)
}

export class NotFoundRoute extends RouteNode {
  constructor(matchAll: boolean, parent: RouteNode | undefined, notFoundComponent: RouteComponent) {
    super(
      {
        path: matchAll ? "*" : "/*",
        component: notFoundComponent,
      },
      parent
    )
  }
}

export const createRouteNodeFromConfig = (routeConfig: RouteConfig[], fns: PipeFn[] = []): RouteNode[] => {
  const _p = (routes: RouteNode[]): RouteNode[] => {
    if (isRealEmpty(fns)) {
      return routes
    }

    // @ts-ignore
    return pipe(...fns)(routes) as RouteNode[]
  }

  return compose<RouteConfig[], RouteNode[], RouteNode[]>(_p, mapRouteNode(undefined))(routeConfig)
}

const ComponentNotDefined: FunctionComponent = () => {
  return <div>Route Component Undefined</div>
}

export interface BuildElementsOptions {
  routeTag?: RouteTagType
  notFound?: RouteComponent
}

/**
 * 获取生成的 Route jsx 标签结构
 */
export const createRouteElements = (
  routeNodes: RouteNode[] = [],
  elementsOptions?: BuildElementsOptions
): JSX.Element => {
  const RouteTag = elementsOptions?.routeTag || Route
  const NotFoundComponent = elementsOptions?.notFound || DefaultNotFoundComponent

  const mapToElements = map<RouteNode, JSX.Element>((routeNode: RouteNode) => {
    const { path, meta, children = [], component: Component } = routeNode

    const routeProps = mergeDeepRight(
      {
        exact: false,
        strict: true,
        sensitive: false,
      },
      routeNode.routeProps || {}
    )

    if (!routeNode.hasChildren) {
      if (meta.redirect) {
        return <Redirect key={path} from={path} to={meta.redirect} exact strict />
      }

      const RenderComponent = Component || ComponentNotDefined

      return <RouteTag {...routeProps} key={path} path={path} component={RenderComponent} />
    }

    return (
      <RouteTag
        {...routeProps}
        key={path}
        path={path}
        render={(props: RouteComponentProps<any>) => {
          const childrenContent = Component ? (
            <Component {...props}>{transformNodeToElement(children, props.location)}</Component>
          ) : (
            transformNodeToElement(children, props.location)
          )

          if (meta.redirect) {
            return (
              <>
                <RouteRedirect routeNode={routeNode} />
                {childrenContent}
              </>
            )
          }

          return (
            <>
              <RouteAutoRedirect routeNode={routeNode} />
              {childrenContent}
            </>
          )
        }}
      />
    )
  })

  const transformNodeToElement = (routeNodes: RouteNode[], location: RouteLocation): JSX.Element => {
    const wrapSwitchTag = (elements: JSX.Element[]) => <Switch location={location}>{elements}</Switch>

    const addNotFound = operators.map((node) => {
      if (node.hasChildren) {
        return node.updateChildren((children) => children?.concat(new NotFoundRoute(false, node, NotFoundComponent)))
      }

      return node
    })

    return compose(wrapSwitchTag, mapToElements, addNotFound)(routeNodes)
  }

  const fullRouteNodes = routeNodes.concat(new NotFoundRoute(false, undefined, NotFoundComponent))

  return transformNodeToElement(fullRouteNodes, undefined)
}

export interface RouterViewProps {
  routes: RouteNode[]
  elements: JSX.Element | null
}

const makeRoutes = (
  routeConfig: RouteConfig[],
  fns: PipeFn[] = [],
  options: BuildElementsOptions = {
    routeTag: Route,
    notFound: DefaultNotFoundComponent,
  }
) => {
  const routes = createRouteNodeFromConfig(routeConfig, fns)
  const elements = createRouteElements(routes, {
    routeTag: options.routeTag,
    notFound: options.notFound,
  })

  return {
    routes,
    elements,
  }
}

export const useRouterView = (getRouteConfig: () => RouteConfig[], fns?: PipeFn[], options?: BuildElementsOptions) => {
  const [routes, setRoutes] = useState<RouteNode[]>([])
  const [elements, setElements] = useState<JSX.Element | null>(null)
  const update = useRef(() => {
    const result = makeRoutes(getRouteConfig(), fns, options)

    setRoutes(result.routes)
    setElements(result.elements)
  })

  const RouteView = memo(() => (
    <RouterContext.Provider
      value={{
        routes: routes,
        elements: elements,
      }}
    >
      {elements}
    </RouterContext.Provider>
  ))

  RouteView.displayName = "RouteView"

  return {
    refresh: update.current,
    View: RouteView,
  }
}
