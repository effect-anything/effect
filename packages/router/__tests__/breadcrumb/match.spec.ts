import { createRouteNodeFromConfig, findRouteNodeByPathname } from "../../src/helper"
import {
  getBreadcrumbs,
  matchRouteNodeBreadcrumb,
  getMetaBreadcrumb,
  breadcrumbRegexPathJoin,
} from "../../src/breadcrumb/match"
import type { RouteConfig } from "../../src/types"
import type { RouteNode } from "../../src"

describe("getBreadcrumbs", () => {
  let routeNodes: RouteNode[] = []

  beforeEach(() => {
    routeNodes = createRouteNodeFromConfig([
      {
        path: "/home",
        meta: {
          breadcrumb: {
            "*": "home",
            "/page1*": [
              {
                text: "page1",
                to: "/home/page1",
              },
            ],
            "/page2*": {
              text: "page2",
              to: "/home/page2",
            },
            "/page3*": () => {
              return [
                {
                  text: "page3",
                  to: "/home/page3",
                },
              ]
            },
          },
        },
        children: [
          {
            path: "/page1",
            meta: {
              breadcrumb: {
                "/nested": {
                  text: "nested",
                  to({ pathname }) {
                    return pathname
                  },
                },
              },
            },
            children: [
              {
                path: "/",
                routeProps: {
                  exact: true,
                },
              },
              {
                path: "/nested",
              },
            ],
          },
          {
            path: "/page2",
            meta: {
              breadcrumb: {
                "/nested": {
                  text: "nested",
                },
              },
            },
            children: [
              {
                path: "/",
                routeProps: {
                  exact: true,
                },
              },
              {
                path: "/nested",
              },
            ],
          },
          {
            path: "/page3",
            meta: {
              breadcrumb: {
                "/nested": {
                  text: "nested",
                },
              },
            },
            children: [
              {
                path: "/",
                routeProps: {
                  exact: true,
                },
              },
              {
                path: "/nested",
              },
            ],
          },
        ],
      },
    ])
  })

  it("should match breadcrumbs result", () => {
    const pathname = "/home"

    const findRouteNodes = findRouteNodeByPathname(pathname, routeNodes)

    expect(findRouteNodes).not.toBeUndefined()

    const result = getBreadcrumbs(
      {
        pathname: pathname,
        query: {},
      },
      findRouteNodes
    )

    expect(result).toEqual([{ text: "home", to: "/home/page1" }])
  })

  it("should match breadcrumbs result2", () => {
    const pathname = "/home/page1"

    const findRouteNodes = findRouteNodeByPathname(pathname, routeNodes)

    expect(findRouteNodes).not.toBeUndefined()

    const result = getBreadcrumbs(
      {
        pathname: pathname,
        query: {},
      },
      findRouteNodes
    )

    expect(result).toEqual([
      { text: "home", to: "/home/page1" },
      { text: "page1", to: "/home/page1" },
    ])
  })

  it("should match breadcrumbs result3", () => {
    const pathname = "/home/page1/nested"

    const findRouteNodes = findRouteNodeByPathname(pathname, routeNodes)

    expect(findRouteNodes).not.toBeUndefined()

    const result = getBreadcrumbs(
      {
        pathname: pathname,
        query: {},
      },
      findRouteNodes
    )

    expect(result).toEqual([
      { text: "home", to: "/home/page1" },
      { text: "page1", to: "/home/page1" },
      { text: "nested", to: "/home/page1/nested" },
    ])
  })

  // xit("should match breadcrumbs empty", () => {
  //   const routes = createRouteNodeFromConfig(config)

  //   const pathname = "/a"

  //   const findRoute = findRouteNodeByPathname(pathname, routes)

  //   const result = getBreadcrumbs(
  //     {
  //       pathname: pathname,
  //       query: {},
  //     },
  //     findRoute
  //   )
  // })
})

describe("matchBreadcrumbNode", () => {
  it("should equal [string]", () => {
    const matchBreadcrumbs = matchRouteNodeBreadcrumb(
      {
        pathname: "/a",
        query: {
          a: 1,
        },
      },
      "/a",
      "/a*",
      "hello"
    )

    expect(matchBreadcrumbs).toEqual([{ text: "hello", to: "/a" }])
  })

  it("should equal [object]", () => {
    const matchBreadcrumbs = matchRouteNodeBreadcrumb(
      {
        pathname: "/a",
        query: {
          a: 1,
        },
      },
      "/a",
      "/a*",
      {
        text: "a",
      }
    )

    expect(matchBreadcrumbs).toEqual([{ text: "a", to: "/a" }])
  })

  it("should equal [objectArray]", () => {
    const matchBreadcrumbs = matchRouteNodeBreadcrumb(
      {
        pathname: "/a",
        query: {
          a: 1,
        },
      },
      "/a",
      "/a*",
      [
        {
          text: "a",
        },
        {
          text: "b",
        },
      ]
    )

    expect(matchBreadcrumbs).toEqual([
      { text: "a", to: "/a" },
      { text: "b", to: "/a" },
    ])
  })

  it("should equal [function]", () => {
    const matchBreadcrumbs = matchRouteNodeBreadcrumb(
      {
        pathname: "/a",
        query: {
          a: 1,
        },
      },
      "/a",
      "/a*",
      () => [
        {
          text: "a",
        },
        {
          text: "b",
        },
      ]
    )

    expect(matchBreadcrumbs).toEqual([
      { text: "a", to: "/a" },
      { text: "b", to: "/a" },
    ])
  })
})

describe("breadcrumbRegexPathJoin", () => {
  it("empty split", () => {
    const p = ["/a", "*"]

    expect(breadcrumbRegexPathJoin(...p)).toEqual("/a*")
  })

  it("too much split", () => {
    const p = ["/a", "//*"]

    expect(breadcrumbRegexPathJoin(...p)).toEqual("/a/*")
  })

  it("pass", () => {
    const p = ["/a", "/b", "/c"]

    expect(breadcrumbRegexPathJoin(...p)).toEqual("/a/b/c")
  })

  it("repeat", () => {
    const p = ["/a/", "//b", "/c", "//d", "/e/"]

    expect(breadcrumbRegexPathJoin(...p)).toEqual("/a/b/c/d/e")
  })
})

describe("getMetaBreadcrumb", () => {
  it("should get meta breadcrumbs", () => {
    const config: RouteConfig[] = [
      {
        path: "/a",
        meta: {
          breadcrumb: {
            "*": "a-0",
            "/1": [
              {
                text: "a-1",
                to: "to a-1",
              },
              {
                text: "a-1-1",
                to: "to a-1-1",
              },
            ],
            "/2": {
              text: "a-2",
              to: "to a-2",
            },
            "/3": () => {
              return [
                {
                  text: "a-3",
                  to: "to a-3",
                },
                {
                  text: "a-3-1",
                  to: "to a-3-1",
                },
              ]
            },
          },
        },
      },
      {
        path: "/a/b",
        meta: {
          breadcrumb: {
            "*": "b-0",
            "/1": [
              {
                text: "b-1",
                to: "to b-1",
              },
              {
                text: "b-1-1",
                to: "to b-1-1",
              },
            ],
            "/2": {
              text: "b-2",
              to: "to b-2",
            },
            "/3": () => {
              return [
                {
                  text: "b-3",
                  to: "to b-3",
                },
                {
                  text: "b-3-1",
                  to: "to b-3-1",
                },
              ]
            },
          },
        },
      },
    ]

    const routes = createRouteNodeFromConfig(config)

    const breadcrumbs = getMetaBreadcrumb(routes)

    expect(breadcrumbs).toMatchSnapshot()
  })
})
