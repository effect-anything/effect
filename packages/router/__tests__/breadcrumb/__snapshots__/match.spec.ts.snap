// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBreadcrumbs should match breadcrumbs children result 1`] = `Array []`;

exports[`getMetaBreadcrumb should get meta breadcrumbs 1`] = `
Array [
  Object {
    "children": undefined,
    "component": undefined,
    "meta": Object {
      "_path": "/a/b",
      "autoRedirectPath": undefined,
      "breadcrumb": Object {
        "*": "b-0",
        "/1": Array [
          Object {
            "text": "b-1",
            "to": "to b-1",
          },
          Object {
            "text": "b-1-1",
            "to": "to b-1-1",
          },
        ],
        "/2": Object {
          "text": "b-2",
          "to": "to b-2",
        },
        "/3": [Function],
      },
    },
    "path": "/a/b",
    "routeProps": Object {},
  },
  Object {
    "children": undefined,
    "component": undefined,
    "meta": Object {
      "_path": "/a",
      "autoRedirectPath": undefined,
      "breadcrumb": Object {
        "*": "a-0",
        "/1": Array [
          Object {
            "text": "a-1",
            "to": "to a-1",
          },
          Object {
            "text": "a-1-1",
            "to": "to a-1-1",
          },
        ],
        "/2": Object {
          "text": "a-2",
          "to": "to a-2",
        },
        "/3": [Function],
      },
    },
    "path": "/a",
    "routeProps": Object {},
  },
]
`;

exports[`getMetaBreadcrumbs should get meta breadcrumbs 1`] = `
Array [
  Object {
    "children": undefined,
    "component": undefined,
    "meta": Object {
      "_path": "/a/b",
      "autoRedirectPath": undefined,
      "breadcrumb": Object {
        "*": "b-0",
        "/1": Array [
          Object {
            "text": "b-1",
            "to": "to b-1",
          },
          Object {
            "text": "b-1-1",
            "to": "to b-1-1",
          },
        ],
        "/2": Object {
          "text": "b-2",
          "to": "to b-2",
        },
        "/3": [Function],
      },
    },
    "path": "/a/b",
    "routeProps": Object {},
  },
  Object {
    "children": undefined,
    "component": undefined,
    "meta": Object {
      "_path": "/a",
      "autoRedirectPath": undefined,
      "breadcrumb": Object {
        "*": "a-0",
        "/1": Array [
          Object {
            "text": "a-1",
            "to": "to a-1",
          },
          Object {
            "text": "a-1-1",
            "to": "to a-1-1",
          },
        ],
        "/2": Object {
          "text": "a-2",
          "to": "to a-2",
        },
        "/3": [Function],
      },
    },
    "path": "/a",
    "routeProps": Object {},
  },
]
`;
